src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs

```diff
@@ -34698,32 +34629,154 @@ .locals init (System.Collections.Generic.List<T> V_0,
+                  .maxstack  1
+                  IL_0000:  ldarg.0
+                  IL_0001:  call       "System.Collections.Generic.List<int> System.Linq.Enumerable.ToList<int>(System.Collections.Generic.IEnumerable<int>)"
+                  IL_0006:  ret
                 }
+                """);
+        }
+
+        [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/71217")]
+        public void List_SingleSpread_IEnumerable_Covariant()
+        {
+            var source = """
+                using System.Collections.Generic;
 
                 class C
                 {
                     static void Main()
                     {
-                        M(new([1, 2, 3])).Report();
+                        M(["a", "b", "c"]).Report();
                     }
 
-                    static List<int> M(MyCollection c) => [..c];
+                    static List<object> M(IEnumerable<string> e) => [..e];
                 }
                 """;
 
-            var verifier = CompileAndVerify([source, s_collectionExtensions], expectedOutput: IncludeExpectedOutput("[1, 2, 3],"), targetFramework: TargetFramework.Net80, verify: Verification.Skipped);
+            var verifier = CompileAndVerify([source, s_collectionExtensions], expectedOutput: "[a, b, c],", verify: Verification.Skipped);
+            verifier.VerifyDiagnostics();
+
+            verifier.VerifyIL("C.M", """
+                {
+                  // Code size        7 (0x7)
+                  .maxstack  1
+                  IL_0000:  ldarg.0
+                  IL_0001:  call       "System.Collections.Generic.List<object> System.Linq.Enumerable.ToList<object>(System.Collections.Generic.IEnumerable<object>)"
+                  IL_0006:  ret
+                }
+                """);
+        }
+
+        [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/71217")]
+        public void List_SingleSpread_IEnumerable_ClassConstraint()
+        {
+            var source = """
+                using System.Collections.Generic;
+                class C
+                {
+                    static void Main()
+                    {
+                        M<IEnumerable<string>, string>(["a", "b", "c"]).Report();
+                    }
+
+                    static List<U> M<T, U>(T t) where T : class, IEnumerable<U> => [..t];
+                }
+                """;
+
+            var verifier = CompileAndVerify([source, s_collectionExtensions], expectedOutput: "[a, b, c],", verify: Verification.Skipped);
+            verifier.VerifyDiagnostics();
+
+            verifier.VerifyIL("C.M<T, U>", """
+                {
+                  // Code size        7 (0x7)
+                  .maxstack  1
+                  IL_0000:  ldarg.0
+                  IL_0001:  call       "System.Collections.Generic.List<U> System.Linq.Enumerable.ToList<U>(System.Collections.Generic.IEnumerable<U>)"
+                  IL_0006:  ret
+                }
+                """);
+        }
```

Comment:
>} [](http://example.com/codeflow?start=8&length=1)

Let's also test:
```
static List<T> M1<T, U>(IEnumerable<U> e) where U : T => [.. e];
static List<T> M2<T, U>(IEnumerable<U> e) where U : struct, T => [.. e];
static List<T> M3<T, U>(IEnumerable<U> e) where U : class, T => [.. e];
static List<T?> M4<T>(IEnumerable<T> e) where T : struct => [.. e];
```

Reply:
Let's test the following:
```csharp
using System.Collections;
using System.Collections.Generic;
class A : IEnumerable<int>
{
    public IEnumerator<int> GetEnumerator() => (new List<int> { 1, 2, 3 }).GetEnumerator();
    IEnumerator<int> IEnumerable<int>.GetEnumerator() => null;
    IEnumerator IEnumerable.GetEnumerator() => null;
}
class C
{
    static void Main()
    {
        M(new A()).Report();
    }
    static List<int> M(A a) => [..a];
}
```

Reply:
The codegen we shipped uses the `IEnumerable<int>.GetEnumerator()` method (via List.AddRange). [SharpLab](https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK7wCYgNQB8ACADAAT4CMAdAMID2ANnQKYDGCUNMAzgNwCwAUEVJkALH0EAmYgEFiIYQGYAPLAQA+AQG8BxXaQXEAkgFEYqMI2QQENZCpjriAcUYJT5y9dsAKAJTEAXjVibxhGAHdiABkoTgR7R01iMjhiCVSDAF9fChc3MwsrG2Q/cT0jd0KvO1Vg8mVa3NdKz2K/QOCzBjK9EwLW20Um/I8in38g4i66cUyBfCkqLR09cgA2UhFiAFloGD8V3W1+cvKAM1tGCGYACxCANwhkYgAPYlgd0IiZP19fHt0cxOqzIGxicQSwW23lkEAmwQA2hQKBAALqzIA=)

IMO, as part of "well-behavedness" of collections, we should require that when `IEnumerable<T>.GetEnumerator()` implementation is a different method than public `GetEnumerator()`, they both yield the same elements from enumeration.

Reply:
> The codegen we shipped uses the `IEnumerable<int>.GetEnumerator()` method (via List.AddRange). [SharpLab](https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK7wCYgNQB8ACADAAT4CMAdAMID2ANnQKYDGCUNMAzgNwCwAUEVJkALH0EAmYgEFiIYQGYAPLAQA+AQG8BxXaQXEAkgFEYqMI2QQENZCpjriAcUYJT5y9dsAKAJTEAXjVibxhGAHdiABkoTgR7R01iMjhiCVSDAF9fChc3MwsrG2Q/cT0jd0KvO1Vg8mVa3NdKz2K/QOCzBjK9EwLW20Um/I8in38g4i66cUyBfCkqLR09cgA2UhFiAFloGD8V3W1+cvKAM1tGCGYACxCANwhkYgAPYlgd0IiZP19fHt0cxOqzIGxicQSwW23lkEAmwQA2hQKBAALqzIA=)

Ah, thanks. I think we should still test this case (although perhaps returning something other than `null` from `IEnumerable<int>.GetEnumerator()`), to capture the actual behavior.